ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/CompactionHeartbeatService.java

```diff
@@ -0,0 +1,256 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.hadoop.hive.ql.txn.compactor;
+
+import org.apache.commons.pool2.BasePooledObjectFactory;
+import org.apache.commons.pool2.ObjectPool;
+import org.apache.commons.pool2.PooledObject;
+import org.apache.commons.pool2.impl.DefaultPooledObject;
+import org.apache.commons.pool2.impl.GenericObjectPool;
+import org.apache.commons.pool2.impl.GenericObjectPoolConfig;
+import org.apache.hadoop.hive.conf.HiveConf;
+import org.apache.hadoop.hive.metastore.HiveMetaStoreUtils;
+import org.apache.hadoop.hive.metastore.IMetaStoreClient;
+import org.apache.hadoop.hive.metastore.conf.MetastoreConf;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.NoSuchElementException;
+import java.util.Objects;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Singleton service responsible for heartbeating the compaction transactions.
+ */
+class CompactionHeartbeatService {
+
+
+  private static final Logger LOG = LoggerFactory.getLogger(CompactionHeartbeatService.class);
+
+  private static volatile CompactionHeartbeatService INSTANCE;
+
+  /**
+   * Return the singleton instance of this class.
+   * @param conf The {@link HiveConf} used to create the service. Used only during the firsst call
+   * @return Returns the singleton {@link CompactionHeartbeatService}
+   * @throws IllegalStateException Thrown when the service has already been shut down.
+   */
+  static CompactionHeartbeatService getInstance(HiveConf conf) {
+    if (INSTANCE == null) {
+      synchronized (CompactionHeartbeatService.class) {
+        if (INSTANCE == null) {
+          LOG.debug("Initializing compaction txn heartbeater service.");
+          INSTANCE = new CompactionHeartbeatService(conf);
+          Runtime.getRuntime().addShutdownHook(new Thread(() -> {
+            try {
+              INSTANCE.shutdown();
+            } catch (InterruptedException e) {
+              Thread.currentThread().interrupt();
+            }
+          }));
+        }
+      }
+    }
+    if (INSTANCE.heartbeatExecutor.isShutdown()) {
+      throw new IllegalStateException("The CompactionHeartbeatService is already shut down!");
+    }
+    return INSTANCE;
+  }
+
+  private final ScheduledExecutorService heartbeatExecutor;
+  private final ObjectPool<IMetaStoreClient> clientPool;
+  private final long initialDelay;
+  private final long period;
+  private final ConcurrentHashMap<Long, TaskWrapper> tasks = new ConcurrentHashMap<>(30);
```

Comment:
why should it be a concurrent map, I don't see any usage of it in the code

Reply:
The map is used in the `startHeartbeat()` and `stopHeartBeat()` methods which can be called concurrently if there a re multiple worker threads configured.

Reply:
and?